### PR TITLE
Initial commit.

### PR DESCRIPTION
**Pull request by athornton**
This seems to be working.

### CODE REVIEW

**Comment by jonathansick** [comment]
Great, I'll take a look by Monday.

**Comment by jonathansick** [comment]
I should add a TEST series to https://github.com/lsst-sqre/lsst-technote-bootstrap so we can try out the UX without creating real technotes.

**Comment by athornton** [comment]
As long as we have delete permissions on the org, it doesn't really matter.

**Comment by jonathansick** [comment]
@athornton Once the technote bootstrap also initializes the technote with keeper.lsst.codes, then it'll also be in that database, and it'd be good to clearly separate test technotes from real technotes so that, for example, we can have a periodic cleanup script.

File: Dockerfile [hunk 1]
```diff
@@ -7,15 +7,16 @@ USER       root
 RUN        yum install -y epel-release
 RUN        yum repolist
 RUN        yum install -y git python-pip python-devel
+RUN	   yum install -y gcc openssl-devel
```
Comment (jhoblitt) [inline Dockerfile]: indenting

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python
+"""Field-calculating plugins for lsst-technote-bootstrap.  For each
+field you want substituted, make a function whose name the field name
+you want to change and which takes a dictionary as input.  It should
+returns the value that field should have after substitution.  Then
+call the plugins substitute() function with the cookiecutter template
+type as the first string argument and your
+dictionary-requiring-substitution as input, and it will change the
+values in that dictionary.
+
```
Comment (jhoblitt) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py]: Is this an intentional newline?

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    """Find the next available serial number for the specified series."""
+    series = inputdict["series"].lower()
+    # Requires that the series pick list has already been replaced with a
+    #  string.
+    # Derive github_org from the series
+    gh_org = ORGHASH[series]
+
+    # scanning the product list won't work.  We need to find the next
+    #  Github repo to use.
+
```
Comment (jhoblitt) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py]: extra newline

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    except github3.exceptions.AuthenticationFailed:
+        raise BackendError(status_code=401,
+                           reason="Bad credentials",
+                           content="Github login failed.")
+    max_serial = 0
+    matchstr = gh_org + "/" + series + "-"
+    for repo in ghub.repositories():
+        rnm = str(repo)
+        if rnm.startswith(matchstr):
+            serstr = rnm[(len(matchstr)):]
```
Comment (jhoblitt) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py:45]: Please comment as to what you want to slice out of the string.

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                sernum = int(serstr)
+                if sernum > max_serial:
+                    max_serial = sernum
+            except ValueError:
+                # We take "couldn't decode" as "not a serial"
+                pass
+    serial = "%03d" % (max_serial + 1)
+    # Actually the same as github_namespace, but the Jinja2 substitution will
+    #  not have happened yet.
+    if "github_repo" not in inputdict or not inputdict["github_repo"]:
```
Comment (jhoblitt) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py]: I just threw up a little...

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        inputdict["github_repo"] = gh_org + "/" + series + "-" + serial
+    return serial
+
+
+def github_org(auth, inputdict):
+    """Derive Github Org from the series."""
+    auth = auth
```
Comment (jhoblitt) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py]: I really dislike this as a linter control idiom.

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+from apikit import BackendError
+from .generic import current_year
+
+ORGHASH = {"sqr": "lsst-sqre",
+           "dmtn": "lsst-dm",
+           "smtn": "lsst-sims",
+           "ldm": "lsst"}
+
+
+def serial_number(auth, inputdict):
```
Comment (jhoblitt) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py:21]: Do you have any Hungarian ancestry? Don't answer that -- HR probably won't let me ask.

File: uservice_ccutter/server.py [hunk 8]
```diff
@@ -1,61 +1,281 @@
...
+                                   content="Github login failed.")
+            namespc = userdict["github_repo"]
+            org, name = namespc.split('/')
+            desc = ""
+            if "description" in userdict:
+                desc = userdict["description"]
+            gh_org = None
+            for gorg in ghub.organizations():
+                if gorg.login == org:
+                    gh_org = gorg
```
Comment (jhoblitt) [inline uservice_ccutter/server.py]: How about `larry, moe, curly` here instead of `gh_org, gorg, org`, it would actually be easier to understand.

**Review by jhoblitt — CHANGES_REQUESTED** [review]
At least the whitespace errors need to be fixed.

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+language: python
+matrix:
+  include:
+  - env: TEST='py.test'
+    python: '2.7'
+    install:
+    - pip install -r requirements.txt
+    - pip install -e .
+    script: py.test
+  - env: TEST='py.test'
```
Comment (jonathansick) [inline .travis.yml:11]: What does the `TEST` variable do?

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    python: '3.5'
+    install:
+    - pip install -r requirements.txt
+    - pip install -e .
+    script: py.test
```
Comment (jonathansick) [inline .travis.yml]: Recommendation: `py.test --flake8 --cov=uservice_ccutter`

Using the `pytest-flake8` and `pytest-cov` packages from PyPI. This is an easy way to make sure PRs conform to PEP8.

File: Dockerfile [hunk 11]
```diff
@@ -7,15 +7,16 @@ USER       root
...
 RUN        useradd -d /home/flasker -m flasker
 RUN        mkdir /dist
 
 # Must run python setup.py sdist first.
 ARG        VERSION="0.0.1"
 LABEL      version="$VERSION"
-COPY       dist/sqre-uservice-ltdstatus-$VERSION.tar.gz /dist
-RUN        pip install /dist/sqre-uservice-ltdstatus-$VERSION.tar.gz
+COPY       dist/sqre-uservice-ccutter-$VERSION.tar.gz /dist
+RUN        pip install /dist/sqre-uservice-ccutter-$VERSION.tar.gz
```
Comment (jonathansick) [inline Dockerfile:19]: So I'm intrigued about this. I've always done `COPY` on the whole application source tree and done a `pip install -r requirements.txt` from inside the image. I guess this is more efficient?

The only worry I have is that someone could change the app version in `setup.py` but forget to update the dockerfile and then wonder why their docker image isn't using the updated codebase. I think it'd be good to have either (or both):

- Some documentation in each microservice, or in appkit or in microservice cookiecutter, that describes how to build and test a microservice.
- Some automation so that the `VERSION` in the dockerfile always follows the version in `setup.py`.

File: README.md [hunk 12]
```diff
@@ -13,8 +14,51 @@ LSST DM SQuaRE api.lsst.codes-compliant microservice wrapper.  TODO
 
 * `/`: returns `OK` (used by Google Container Engine Ingress healthcheck)
 
-* `/ccutter`: TODO
+* `/ccutter`: returns a JSON structure.  The keys are the types of
+  projects the service knows how to make with cookiecutter, and the values
+  are what cookiecutter expects to use as cookiecutter.json for that
+  type of project.
+  
+* `/ccutter/<projecttype>`: returns a JSON structure for the named
```
Comment (jonathansick) [inline README.md]: Might be good to have a section in the README that lists the implemented `projecttype`s.

File: README.md [hunk 13]
```diff
@@ -13,8 +14,51 @@ LSST DM SQuaRE api.lsst.codes-compliant microservice wrapper.  TODO
...
+  project type.
+  
+* `/ccutter/<projecttype>` (POST): accepts JSON representing a
+  filled-out cookiecutter.json form.  This must be authenticated: the
+  authentication headers should contain a username of the Github user
```
Comment (jonathansick) [inline README.md]: Github -> GitHub (search/replace project wide)

File: README.md [hunk 14]
```diff
@@ -13,8 +14,51 @@ LSST DM SQuaRE api.lsst.codes-compliant microservice wrapper.  TODO
...
+    * Substitutes additional fields in the JSON depending on the project
+      type.
+    * Runs cookiecutter to create the project from the template.
+	* Creates a repository on Github for the project.
+	* Pushes the project content to Github
+  It returns a JSON structure with one field "github_repo", which
+  contains the HTTPS clone url of the new repository.
 
-### Returned Structure
+### Adding new project types
```
Comment (jonathansick) [inline README.md:40]: Excellent 👍

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+                if sernum > max_serial:
+                    max_serial = sernum
+            except ValueError:
+                # We take "couldn't decode" as "not a serial"
+                pass
+    serial = "%03d" % (max_serial + 1)
+    # Actually the same as github_namespace, but the Jinja2 substitution will
+    #  not have happened yet.
+    if "github_repo" not in inputdict or not inputdict["github_repo"]:
+        inputdict["github_repo"] = gh_org + "/" + series + "-" + serial
```
Comment (jonathansick) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py]: I might not be following properly, but for technotes I'd rather not ever get the `github_repo` from the user; I'd always want it calculated from the series and serial number. I.e., the only input a user should provide is `series` to compute `gh_org`, `serial`, `github_repo`.

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py (hunk 6 above)
Comment (jonathansick) [reply uservice_ccutter/plugins/lsst_technote_bootstrap.py]: I agree with @jhoblitt here. DM decided not to user pylint as our linter since it neither implements PEP8, nor do the recommendations always lead to cleaner code.

My read on pylint is that it's fine to run locally on your code before making a PR to check if there are obvious things to clean up, but it's counter-productive to use as a live linter in your editor. pyflakes is a much better linter for that because *pyflakes is always correct* (i.e., that's what we use in CI to block merges).

I'd say it's not even worth putting extra pylint control comments in code. pylint isn't part of the DM standard. It's definitely bad form to obfuscate code with `auth = auth`.

File: uservice_ccutter/projecturls.py [hunk 16]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline uservice_ccutter/projecturls.py]: not a script

File: uservice_ccutter/plugins/uservice_bootstrap.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline uservice_ccutter/plugins/uservice_bootstrap.py]: not a script

File: uservice_ccutter/plugins/substitute.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline uservice_ccutter/plugins/substitute.py]: not a script

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline uservice_ccutter/plugins/lsst_technote_bootstrap.py]: not a script

File: uservice_ccutter/plugins/generic.py [hunk 20]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline uservice_ccutter/plugins/generic.py]: not a script

File: uservice_ccutter/plugins/__init__.py [hunk 21]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env/python
```
Comment (jonathansick) [inline uservice_ccutter/plugins/__init__.py]: not a script

File: uservice_ccutter/__init__.py [hunk 22]
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env python
```
Comment (jonathansick) [inline uservice_ccutter/__init__.py]: not a script

File: tests/test_substitute_uservice.py [hunk 23]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline tests/test_substitute_uservice.py:1]: not a script

File: tests/test_substitute_technote.py [hunk 24]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
```
Comment (jonathansick) [inline tests/test_substitute_technote.py:1]: not a script

File: Dockerfile [hunk 25]
```diff
@@ -7,15 +7,16 @@ USER       root
 RUN        yum install -y epel-release
 RUN        yum repolist
 RUN        yum install -y git python-pip python-devel
+RUN	   yum install -y gcc openssl-devel
 RUN        pip install --upgrade pip
 RUN        useradd -d /home/flasker -m flasker
```
Comment (jonathansick) [inline Dockerfile:12]: Maybe use a more relevant username than the name of the demo flask project.

File: README.md [hunk 26]
```diff
@@ -13,8 +14,51 @@ LSST DM SQuaRE api.lsst.codes-compliant microservice wrapper.  TODO
 
 * `/`: returns `OK` (used by Google Container Engine Ingress healthcheck)
```
Comment (jonathansick) [inline README.md]: It's a good idea to always prefix the endpoint with the method (I know you do this for `/ccutter/<projecttype`, but I'd like to see it for all. E.g.:

`GET /`: return `'OK'` [..]

File: uservice_ccutter/server.py [hunk 27]
```diff
@@ -1,61 +1,281 @@
...
+        @app.route("/ccutter/<ptype>/", methods=["GET", "POST"])
+        # pylint: disable=unused-variable
+        def action_for_type(ptype):
+            """Either return the template, or create a new thing"""
+            if request.method == "GET":
+                return jsonify(get_single_project_type(ptype))
+            return post_request(ptype)
+
+        def post_request(ptype):
+            """Crete project: write to local repo and to GH."""
```
Comment (jonathansick) [inline uservice_ccutter/server.py]: Create

File: uservice_ccutter/server.py [hunk 28]
```diff
@@ -1,61 +1,281 @@
...
+            app.run(host='0.0.0.0', threaded=True)
 
 
 def standalone():
     """Entry point for running as its own executable."""
     server(run_standalone=True)
 
 
+def _refresh_cache(app, temp_dir, timeout):
+    """Refresh cookiecutter.json cache if needed."""
```
Comment (jonathansick) [inline uservice_ccutter/server.py:245]: I feel like a complete docstring is warranted here. Or at least document the caching behavior somewhere. There's a lot of logic in this function that's pretty hard to follow.

**Review by jonathansick — COMMENTED** [review]
Besides the line comments, I also want to make sure that:

- `series` is always all-caps in the resolved cookiecutter son.
- `repo_name` and `github_namespace` are always computed only from the `series` and not directly set-able by the user.

If `lsst-technote-bootstrap/cookiecutter.json` needs to be re-ordered to simplify the code that can be done too.

**Comment by jonathansick** [comment]
Oh, also the technote cookiecutter part needs to

- Enable Travis for the repo.
- Add encrypted LTD Keeper and S3 credentials to the `.travis.yml` (see LTD Keeper Admin item in 1Password). Note that run-down uses the `travis` rubygem to do the encryption, but I think there's a python package too.
- Needs to register the technote with `keeper.lsst.codes`
- Protect the GitHub `master` branch and require that Travis passes before permitting a merge.

We can do this now or in a separate ticket. When LTD Keeper interaction is involved I'd like to add a new TEST series in lsst-technote-bootstrap so it's easy to find and garbage collect test projects.

**Comment by athornton** [comment]
The LTD Keeper and Travis stuff will definitely be a separate ticket.  I'll get to the rest once I'm back from the sqrbot front-end.

Re: Josh's throwing up a little: is there a better way to find what the next serial is than to look for all projects in the right org named sqr-something, try to parse something as a number, and then take the next number that's bigger than those?  I feel like the real problem here is that we're relying on a fragile and not-really-documented naming convention, but I don't know of any other way to tell, when looking at Github, what is and is not supposed to be a technote.

**Comment by jonathansick** [comment]
> The LTD Keeper and Travis stuff will definitely be a separate ticket. I'll get to the rest once I'm back from the sqrbot front-end.

Totally fine.

> I feel like the real problem here is that we're relying on a fragile and not-really-documented naming convention

Yeah, the naming convention is really the only source of truth. If LTD Keeper provided an API for this it'd still only a proxy to the truth that exists in the GitHub organization.

Is Josh's issue just with the implementation of the code?

**Comment by athornton** [comment]
Trying to determine what the vomitory offense was.  I can't do the serial number extraction as a list comprehension since I may throw an exception trying to decode a not-translatable-to-an-int; once I had a list of serials in use I could use max on the list, although I guess the better answer would be to compare to non-negative integers and use the first one that is in those and isn't in my serial list, so we could backfill missing ones.

Maybe that was it.

File: .travis.yml (hunk 9 above)
Comment (athornton) [reply .travis.yml:11]: I don't know.  .travis.yml is completely cargo-culted.

File: Dockerfile (hunk 25 above)
Comment (athornton) [reply Dockerfile:12]: Not worth fixing at this point.  I need to take all the microservices and make them happy under uwsgi, not running standalone.  When I have a Docker recipe that does that well, then we can make uwsgi run as something sane.

File: Dockerfile (hunk 11 above)
Comment (athornton) [reply Dockerfile:19]: Initially VERSION is set from cookiecutter.  It might be nice to have setup/Dockerfile/the version in server.py all follow the same thing, but I haven't worked through a nice way to get all three of them.  You can do server.py tracking setup with setuptools, but propagating to Dockerfile...well, Dockerfile can't run-a-thing-and-include-its-output, can it?

Future enhancement.  It starts at the same place.  I can add some notes to the cookiecutter docs that says "as you evolve this, you need to change the app version in three places: x, y, z"

File: README.md (hunk 26 above)
Comment (athornton) [reply README.md]: Enhancement to cookiecutter template.

File: uservice_ccutter/plugins/lsst_technote_bootstrap.py (hunk 6 above)
Comment (athornton) [reply uservice_ccutter/plugins/lsst_technote_bootstrap.py]: This is why I am continuing to use pylint:

```(cdk2) adam@ixitxachitl:~/git/sqre-uservice-ccutter/uservice_ccutter$ pyflakes server.py
(cdk2) adam@ixitxachitl:~/git/sqre-uservice-ccutter/uservice_ccutter$ flake8 server.py
(cdk2) adam@ixitxachitl:~/git/sqre-uservice-ccutter/uservice_ccutter$ pylint server.py
No config file found, using default configuration
************* Module uservice_ccutter.server
I: 49, 0: Locally disabling unused-variable (W0612) (locally-disabled)
I: 57, 0: Locally disabling unused-variable (W0612) (locally-disabled)
I: 64, 0: Locally disabling unused-variable (W0612) (locally-disabled)
I: 74, 0: Locally disabling unused-variable (W0612) (locally-disabled)
I:104, 0: Locally disabling too-many-locals (R0914) (locally-disabled)
I:246, 0: Locally disabling too-many-locals (R0914) (locally-disabled)
W: 29,33: Unused argument 'debug' (unused-argument) 
```

Neither pyflakes nor flake8 caught the unused argument there.

I am OK with disabling pylint warnings rather than doing, say, pointless variable assignments to shut it up.  And I am 100% OK with ensuring that my code is clean under pyflakes and flake8 as well as pylint.  But I'm not OK with not using it, since it catches legitimate bugs that the other two don't.  It's fine with me if this remains an idiosyncratic choice of mine, but I find it a helpful tool for me.

I've added flymake/pyflakes to Emacs; that plus py-autopep8 seem to do a pretty good job of getting the clode syntactical and clean.

**Review by jhoblitt — APPROVED** [review]
